src/unicode/mod.rs

```diff
@@ -0,0 +1,185 @@
+use std::cmp::Ordering;
+use std::hash::{Hash, Hasher};
+
+use self::map::lookup;
+mod map;
+
+#[derive(Clone, Copy, Debug)]
+#[cfg_attr(feature = "heap_size", derive(HeapSizeOf))]
+pub struct Unicode<S>(pub S);
+
+impl<S1: AsRef<str>, S2: AsRef<str>> PartialEq<Unicode<S2>> for Unicode<S1> {
+    #[inline]
+    fn eq(&self, other: &Unicode<S2>) -> bool {
+        self.0.as_ref().chars().flat_map(lookup)
+            .zip(other.0.as_ref().chars().flat_map(lookup))
+            .all(|(a, b)| a == b)
+    }
+}
+
+impl<S: AsRef<str>> Eq for Unicode<S> {}
+
+#[cfg(iter_cmp)]
+impl<T: AsRef<str>> PartialOrd for Unicode<T> {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[cfg(iter_cmp)]
+impl<T: AsRef<str>> Ord for Unicode<T> {
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        let self_chars = self.0.as_ref().chars().flat_map(lookup);
+        let other_chars = other.0.as_ref().chars().flat_map(lookup);
+        self_chars.cmp(other_chars)
+    }
+}
+
+impl<S: AsRef<str>> Hash for Unicode<S> {
+    #[inline]
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
```

Comment:
Yep, the test `test_eq_unicode` includes a hash check. Though it's only a single comparison.